ssl/statem/statem_clnt.c

```diff
@@ -1859,9 +1859,10 @@ WORK_STATE tls_post_process_server_certificate(SSL *s, WORK_STATE wst)
     size_t certidx;
     int i;
 
+    if (s->rwstate == SSL_RETRY_VERIFY) /* reset the SSL_RETRY_VERIFY state */
+        s->rwstate = SSL_NOTHING;
     i = ssl_verify_cert_chain(s, s->session->peer_chain);
-    if (i == -1) {
-        s->rwstate = SSL_RETRY_VERIFY;
+    if (i > 0 && s->rwstate == SSL_RETRY_VERIFY) {
```

Comment:
Hmm. The `i > 0` implies hat the retry will be done only on "success". I suggest `i >= 0` here.
See also my related comment on the documentation above.

Reply:
No, I think it should be done only on success.